pkg/loadbalancer/experimental/services_test.go

```diff
@@ -0,0 +1,411 @@
+		for _, addr := range []loadbalancer.L3n4Addr{beAddr1, beAddr2, beAddr3} {
+			be, _, found := p.BackendTable.Get(txn, experimental.BackendAddrIndex.Query(addr))
+			if assert.True(t, found, "Backend not found with address %s", addr) {
+				assert.True(t, be.L3n4Addr.DeepEqual(&addr), "Backend address %s does not match %s", be.L3n4Addr, addr)
+			}
+		}
+
+		// By service
+		bes := statedb.Collect(p.BackendTable.List(txn, experimental.BackendServiceIndex.Query(name1)))
+		require.Len(t, bes, 2)
+		require.True(t, bes[0].L3n4Addr.DeepEqual(&beAddr1))
+		require.True(t, bes[1].L3n4Addr.DeepEqual(&beAddr2))
+
+		// Backends for [name2] can be found even though the service doesn't exist (yet).
+		bes = statedb.Collect(p.BackendTable.List(txn, experimental.BackendServiceIndex.Query(name2)))
+		require.Len(t, bes, 1)
+		require.True(t, bes[0].L3n4Addr.DeepEqual(&beAddr3))
+	}
+
+	// GetBackendsForService
+	{
+		txn := p.DB.ReadTxn()
+
+		bes := statedb.Collect(experimental.GetBackendsForFrontend(txn, p.BackendTable, svc))
+		require.Len(t, bes, 2)
+		require.True(t, bes[0].L3n4Addr.DeepEqual(&beAddr1))
+		require.True(t, bes[1].L3n4Addr.DeepEqual(&beAddr2))
+	}
+
+	// UpdateBackendState
+	{
+		wtxn := p.Services.WriteTxn()
+		err := p.Services.UpdateBackendState(wtxn, beAddr2, loadbalancer.BackendStateMaintenance)
+		require.NoError(t, err, "UpdateBackendState")
+
+		// Service has been bumped and is pending.
+		svc, _, found := p.FrontendTable.Get(wtxn, experimental.FrontendL3n4AddrIndex.Query(frontend))
+		require.True(t, found, "Lookup service")
+		require.Equal(t, p.BackendTable.Revision(wtxn), svc.BackendRevision, "Unexpected backend revision")
+		require.Equal(t, reconciler.StatusKindPending, svc.Status.Kind, "Expected to be pending")
+
+		wtxn.Abort()
+	}
+
+	// ReleaseBackend
+	{
+		wtxn := p.Services.WriteTxn()
+
+		// Release the [name1] reference to [beAddr1].
+		require.Equal(t, 3, p.BackendTable.NumObjects(wtxn))
+		err := p.Services.ReleaseBackend(wtxn, name1, beAddr1)
+		require.NoError(t, err, "ReleaseBackend")
+
+		// [beAddr2] remains for [name1].
+		bes := statedb.Collect(experimental.GetBackendsForFrontend(wtxn, p.BackendTable, svc))
+		require.Len(t, bes, 1)
+		require.True(t, bes[0].L3n4Addr.DeepEqual(&beAddr2))
+
+		wtxn.Abort()
+	}
+
+	// DeleteBackendsBySource
+	{
+		wtxn := p.Services.WriteTxn()
+
+		require.Equal(t, 3, p.BackendTable.NumObjects(wtxn))
+		err := p.Services.DeleteBackendsBySource(wtxn, source.Kubernetes)
+		require.NoError(t, err, "DeleteBackendsBySource")
+		iter, _ := p.BackendTable.All(wtxn)
+		require.Len(t, statedb.Collect(iter), 0)
+
+		// No backends remain for the service.
+		bes := statedb.Collect(experimental.GetBackendsForFrontend(wtxn, p.BackendTable, svc))
+		require.Len(t, bes, 0)
+
+		wtxn.Abort()
+	}
+}
+
+// TestFrontendJSONRoundtrip validates that Frontend can be marshalled and unmarshalled
+// into JSON without information loss. The JSON representation is used for example with the
```

Comment:
Sorry for nitpicking, it's my mathematical side speaking here😅 Forgetting about the `TableRow` check at the end for a sec, since the test only compares JSON bytes, it verifies that marshalling is kind of an "idempotent" in the sense that the sequence `marshall`-`unmarshall`-`marshall` does the same as just `marshall`, but theoretically such a test would pass even if marshalling wiped out all the data.